File: Problem 20 - Valid Parentheses/code.py
class Solution(object):
    def isValid(self, s):
        stack = []
        matching_brackets = {
            ')': '(',
            '}': '{',
            ']': '['
        }

        for char in s:
            if char in "({[":
                stack.append(char)
            else:
                if len(stack) == 0 or stack.pop() != matching_brackets[char]:
                    return False
        
        return len(stack) == 0

    # Case 1 
    # Input: s = "()"
    # Expected Output: True
    print(isValid("", "()"))

    # Case 2
    # Input: s = "()[]{}"
    # Expected Output: True
    print(isValid("", "()[]{}"))

    # Case 3
    # Input: s = "(]"
    # Expected Output: False
    print(isValid("", "(]"))

File: Problem 20 - Valid Parentheses/code.js
var isValid = function(s) {
    const stack = [];
    const matchingBrackets = {
        ')': '(',
        '}': '{',
        ']': '['
    };

    for (let char of s) {
        if (char === '(' || char === '{' || char === '[') {
            stack.push(char);
        } else {
            if (stack.length === 0 || stack.pop() !== matchingBrackets[char]) {
                return false;
            }
        }
    }

    return stack.length === 0;
}

// Case 1 
// Input: s = "()"
// Expected Output: true
console.log(isValid("()"))

// Case 2
// Input: s = "()[]{}"
// Expected Output: true
console.log(isValid("()[]{}"))

// Case 3
// Input: s = "(]"
// Expected Output: false
console.log(isValid("(]"))